src/tests/profiler/native/gcheapenumerationprofiler/gcheapenumerationprofiler.cpp

```diff
@@ -0,0 +1,261 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "gcheapenumerationprofiler.h"
+#include "../profilerstring.h"
+
+#if WIN32
+#define EXPORT
+#else // WIN32
+#define EXPORT __attribute__ ((visibility ("default")))
+#endif // WIN32
+
+GUID GCHeapEnumerationProfiler::GetClsid()
+{
+    // {8753F0E1-6D6D-4329-B8E1-334918869C15}
+	GUID clsid = { 0x8753f0e1, 0x6d6d, 0x4329,{ 0xb8, 0xe1, 0x33, 0x49, 0x18, 0x86, 0x9c, 0x15 } };
+	return clsid;
+}
+
+HRESULT GCHeapEnumerationProfiler::Initialize(IUnknown* pICorProfilerInfoUnk)
+{
+    Profiler::Initialize(pICorProfilerInfoUnk);
+    printf("GCHeapEnumerationProfiler::Initialize\n");
+
+    int eventMask = 0;
+    constexpr ULONG bufferSize = 1024;
+    ULONG envVarLen = 0;
+    WCHAR envVar[bufferSize];
+    HRESULT hr = S_OK;
+
+    if (FAILED(hr = pCorProfilerInfo->GetEnvironmentVariable(WCHAR("Set_Monitor_GC_Event_Mask"), bufferSize, &envVarLen, envVar)))
+    {
+        printf("FAIL: ICorProfilerInfo::GetEnvironmentVariable() failed hr=0x%x", hr);
+        IncrementFailures();
+        return hr;
+    }
+    String envVarStr = envVar;
+    String expectedEnvVarStr = reinterpret_cast<const WCHAR*>(u"TRUE");
+    if (envVarStr == expectedEnvVarStr)
+    {
+        printf("Setting GarbageCollectionStarted event masks\n");
+        eventMask |= COR_PRF_MONITOR_GC;
+    }
+
+    if (FAILED(hr = pCorProfilerInfo->SetEventMask2(eventMask, COR_PRF_HIGH_MONITOR_NONE)))
+    {
+        printf("FAIL: ICorProfilerInfo::SetEventMask2() failed hr=0x%x", hr);
+        IncrementFailures();
+    }
+
+    return hr;
+}
+
+void GCHeapEnumerationProfiler::IncrementFailures()
+{
+    _failures.fetch_add(1, std::memory_order_relaxed);
+}
+
+HRESULT STDMETHODCALLTYPE GCHeapEnumerationProfiler::GarbageCollectionStarted(int cGenerations, BOOL generationCollected[], COR_PRF_GC_REASON reason)
+{
+    SHUTDOWNGUARD();
+    printf("GCHeapEnumerationProfiler::GarbageCollectionStarted\n");
```

Comment:
Is the choice to call EnumerateGCHeapObjects from GarbageCollectionStarted simply an easy way to test? Or can we call it at any point as long as we call SuspendRuntime first?

Reply:
EnumerateGCHeapObjects will handle EE suspension, and as it is an asynchronous API, it can be both directly called by the profiler, as well as invoked by a callback (demonstrated here). The `EnumerateGCHeapObjectsWithoutProfilerRequestedRuntimeSuspension` PInvoke demonstrates invoking the API directly from the profiler.

Calling EnumerateGCHeapObjects from GarbageCollectionStarted was simply to test the behavior when a different native thread has suspended EE (EnumerateGCHeapObjectsMultiThreadWithCompetingRuntimeSuspension test case). In this case, the current thread has suspended EE through GC, and it spins off a separate native thread to EnumerateGCHeapObjects, demonstrating that ThreadSuspend::SuspendEE in the other thread will block until the native thread that owns suspension has RestartedEE.